### PR TITLE
Prefix

### PR DESCRIPTION
**Pull request by ekinimo**
It definetly needs more tests but with in this branch  we can parse and evaluate  prefixes division  subtraction .

### CODE REVIEW

File: src/infrastructure/parser.rs [hunk 1]
```diff
@@ -96,12 +142,15 @@ impl SplitStringParser {
                 let atom = match (int_parse_result, float_parse_result) {
                     (Ok(value), _) => Integer(value),
                     (_, Ok(value)) => Float(value),
-                    (_, _) => return Err("Parse error. This should never happen".to_string())
+                    (_, _) => return Err("Parse error. This should never happen".to_string()),
```
Comment (pubrrr) [inline src/infrastructure/parser.rs:145]: Can we have a more meaningful error message here? What has to happen to get here?

- [ ] to do

**Review by pubrrr — APPROVED** [review]

File: src/infrastructure/lexer.rs [hunk 2]
```diff
@@ -15,6 +15,12 @@ impl StringLexer {
     }
 }
 
+impl Default for StringLexer {
```
[the remark this answers is not in the document]
Comment (ekinimo) [reply src/infrastructure/lexer.rs:18]: clippy suggested it i followed it. it provides default constructor

File: src/infrastructure/parser.rs [hunk 3]
```diff
@@ -29,23 +39,29 @@ impl ExpressionParser<Vec<Token>> for SplitStringParser {
...
+     $left_parser:expr,$right_parser:expr) => {
+        fn $fun_name(mut tokens: &mut Vec<Token>) -> Result<Expression, String> {
+            //println!(" {} {:?}",stringify!($fun_name) ,tokens);
+            let mut left_parseresult = $left_parser(&mut *tokens)?;
+
+            let mut i = 0;
+            loop{
+                println!(" {} loops {} ",stringify!($fun_name),i );
+                i = i+1;
+             left_parseresult = match tokens.last() {
```
[the remark this answers is not in the document]
Comment (ekinimo) [reply src/infrastructure/parser.rs:59]: im not formatting the whole macro stringify turns the macro capture $fun_name into a string

File: src/infrastructure/parser.rs [hunk 4]
```diff
@@ -29,23 +39,29 @@ impl ExpressionParser<Vec<Token>> for SplitStringParser {
...
+                $(Some($token_name) => {
                     tokens.pop();
-                    $right_parser(&mut *tokens)?
-                },
+                    Expression::BinaryOperator($ast_member(
+                            Box::new(left_parseresult),
+                            Box::new($right_parser(&mut tokens)?),
+                        ))
+                },)+
```
[the remark this answers is not in the document]
Comment (ekinimo) [reply src/infrastructure/parser.rs:66]: + is for repetition just like in the regular expressions $($inner_stuff_to_repeat)+ means one or more time

File: src/infrastructure/parser.rs (hunk 1 above)
Comment (ekinimo) [reply src/infrastructure/parser.rs:154]: i believe if we try to parse something like 1231231231231231231231231 that would do the trick. if the number we supplied doesnt fit the i64 or f64 it should throw this error.

File: src/infrastructure/parser.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ekinimo) [reply src/infrastructure/parser.rs:59]: oh formatting in the macro... I understood totally wrong.  well sometimes i forget cargo fmt i should probably bind it to emacs whenever i save a rust file that it should run it

File: src/infrastructure/parser.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ekinimo) [reply src/infrastructure/parser.rs:66]: essentially yes

**Pull request merged**